inject/src/main/java/io/micronaut/context/AbstractBeanDefinition2.java

```diff
@@ -0,0 +1,1846 @@
+import io.micronaut.inject.annotation.AbstractEnvironmentAnnotationMetadata;
+import io.micronaut.inject.qualifiers.InterceptorBindingQualifier;
+import io.micronaut.inject.qualifiers.Qualifiers;
+import io.micronaut.inject.qualifiers.TypeAnnotationQualifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+/**
+ * <p>Default implementation of the {@link BeanDefinition} interface. This class is generally not used directly in user
+ * code.
+ * Instead a build time tool does analysis of source code and dynamically produces subclasses of this class containing
+ * information about the available injection points for a given class.</p>
+ * <p>
+ * <p>For technical reasons the class has to be marked as public, but is regarded as internal and should be used by
+ * compiler tools and plugins (such as AST transformation frameworks)</p>
+ * <p>
+ * <p>The {@link io.micronaut.inject.writer.BeanDefinitionWriter} class can be used to produce bean definitions at
+ * compile or runtime</p>
+ *
+ * @param <T> The Bean definition type
+ * @author Graeme Rocher
+ * @author Denis Stepanov
+ * @see io.micronaut.inject.writer.BeanDefinitionWriter
+ * @since 3.0
+ */
+@Internal
+public class AbstractBeanDefinition2<T> extends AbstractBeanContextConditional implements BeanDefinition<T>, EnvironmentConfigurable {
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractBeanDefinition2.class);
+    private static final String NAMED_ATTRIBUTE = Named.class.getName();
+
+    private final Class<T> type;
+    private final AnnotationMetadata annotationMetadata;
+    private final Optional<String> scope;
+    private final boolean isProvided;
+    private final boolean isIterable;
+    private final boolean isSingleton;
+    private final boolean isPrimary;
+    private final boolean isAbstract;
+    private final boolean isConfigurationProperties;
+    private final boolean hasExposedTypes;
+    private final boolean requiresMethodProcessing;
+    @Nullable
+    private final MethodOrFieldReference constructor;
+    @Nullable
+    private final MethodReference[] methodInjection;
+    @Nullable
+    private final FieldReference[] fieldInjection;
+    @Nullable
+    private final ExecutableMethodsDefinition<T> executableMethodsDefinition;
+    @Nullable
+    private final Map<String, Argument<?>[]> typeArgumentsMap;
+    @Nullable
+    private Environment environment;
+    private Set<Class<?>> exposedTypes;
+    private Optional<Argument<?>> containerElement;
```

Comment:
Same here do we need to store the optoinal? Would it not be sufficient to just make this Nullable?